### ISSUE TITLE
Manual WebRoot Authentication

### ISSUE THREAD

**Issue by iamacarpet** [letsencrypt/letsencrypt#1475]
Hello,

I've been looking at enrolling some of our company websites for LetEncrypt certificates as part of the BETA and we had a slight special use case.

We have compliance requirements which mean we:
* Can't allow the LetsEncrypt software to run on the reverse proxy nodes.
* Can't run anything on port 80/443 other than nginx serving our sites, availability is quite important.
* Need replication to a group of servers before validation, it could pick any one of them.

My solution was to combine the manual and webroot plugins, gist of my change to webroot.py here: https://gist.github.com/iamacarpet/fb4a8f424390c0352e93

To use, you simply:
* Run letsencrypt with the webroot plugin.
* It'll tell you when it's written the files and ask you to install them.
* Copy from the local webroot folder to your remote server(s).
* Make sure web server is configured - will include nginx config below.
* Hit enter back at the letencrypt script & it will validate.
* Install certs to remote server(s) as required.

nginx config:
`location /.well-known/acme-challenge {
    alias /var/data/letsencrypt/.well-known/acme-challenge;
    location ~ /.well-known/acme-challenge/(.*) {
        add_header Content-Type application/jose+json;
    }
}`
Borrowed from here: https://community.letsencrypt.org/t/using-the-webroot-domain-verification-method/1445/6

I feel it would be good as it's own plugin as editing the webroot plugin isn't great, but I'm afraid we don't have the time to complete at the moment, so thought I'd share in it's current state :).

Regards,
IAmACarpet

**Comment by My1** [comment]
wasnt the header changed to text/plain?

**Comment by chriscroome** [comment]
This is a very simple solution for doing domain verification with Let's Encrypt and multiple web servers with round-robin DNS, I think it's great.

**Comment by iamacarpet** [comment]
@My1 Yes only discovered that after I'd already posted, will edit, thank you.

It did still seemed to work ok though.

**Comment by centminmod** [comment]
> I feel it would be good as it's own plugin as editing the webroot plugin isn't great

maybe a separate `webroot-manual` authentication plugin ?

**Comment by kuba** [comment]
Why not use `webroot` and `sshfs` or any other network fs, @iamacarpet?

Alternatively, `manual` plugin gives you exact contents and path to a file that you can just copy and paste...

**Comment by iamacarpet** [comment]
@kuba In a PCI compliant environment, direct access between the box the keys are generated/stored on and the reverse proxy that needs the certificate and host the website isn't always possible - especially in a cluster environment.

The manual plugin gives you a python script (unless I'm mistaken?) which has to be run as root, as well as taking over either port 80 or 443, which means the websites you're hosting would briefly have to be taken down, which in a business environment is again, not really ideal.

It's not for everybody, but this killed our two birds with one stone.

**Comment by My1** [comment]
manual mode gives you 2 options: either place a file at an HTTP accessible location http://example.com/.well-known/acme-challenge/some-code
and the file has to conatin some more text you are given by the LE server. so no restart of your webserver that way.

**Comment by stonie08** [comment]
@My1 the problem with the manual mode is that it doesn't work with a HTTPS-only accessible location (eg. you're using Port 80 for something different and are only running a webserver on port 443)

**Comment by My1** [comment]
okay well then it is annoying...

**Comment by bmw** [comment]
Per the [IETF draft](https://tools.ietf.org/html/draft-ietf-acme-acme-01#section-7.2) for the ACME spec, accessing the file over HTTPS isn't allowed. To quote the relevant sentence:

>Because many webservers allocate a default HTTPS virtual host to a particular low-privilege tenant user in a subtle and non-intuitive manner, the challenge must be completed over HTTP, not HTTPS.

You can see more discussion of this issue [here](https://groups.google.com/a/letsencrypt.org/forum/#!topic/client-dev/xnNL-16VQiU).

**Comment by My1** [comment]
that's why maybe specifying a different port should be made, also the intresting question is what happens as per the IETF when you get a redirect from the HTTP to an HTTPS resource?

**Comment by iveoles** [comment]
Specifying a port would be ideal. We also run in a cluster set up, but I will be generating the certificates on the same server each time. If we could specify the port for the challenge to be completed on, then the rest of the automation takes care of itself.

**Comment by frimelle** [comment]
+1 We can't take down our cluster aswell. Please consider an option to specify a port.

**Comment by atleta** [comment]
I think the right way  to fix this issue would be to make the documentation more clear. It's not obvious (at first sight) why you need to run the python server, esp. if you're not a python developer.

The help text printed by manual mode could also be more clear about it. (Or maybe even add a switch that lets the user specify wether they have  a web server or not, and/or the webroot of the web server, so that the commands that need to be run can be just copy&pasted.)

**Comment by My1** [comment]
I think the challenge should just contain the public key(s) that are allowed to sign for the domain. would make a lot of SANs more easy.

**Comment by bmw** [comment]
We're currently working on making the `manual` plugging scriptable meaning you could have it deploy the challenge files to all of your webservers before asking for verification from the Let's Encrypt CA (see #2782). I believe this should help this use case significantly.

Using a special port is outside of the scope of the client though. Feel free to open an issue/PR against the spec at https://github.com/ietf-wg-acme/acme.
